File: packages/kotlin/src/android/src/main/java/alphaTab/core/ecmaScript/PromiseWithResolvers.kt
package alphaTab.core.ecmaScript

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred

class PromiseWithResolvers<T> {
    private val _deferred = CompletableDeferred<T>()

    public val promise: Deferred<T>
        get() = _deferred

    fun resolve(v: T) {
        _deferred.complete(v)
    }

    fun reject(e: Error) {
        _deferred.completeExceptionally(e)
    }
}

File: packages/csharp/src/AlphaTab/Core/EcmaScript/PromiseWithResolvers.cs
using System.Threading.Tasks;

namespace AlphaTab.Core.EcmaScript;

public class PromiseWithResolvers<T>
{
    private readonly TaskCompletionSource<T> _taskCompletionSource = new();
    public Task<T> Promise => _taskCompletionSource.Task;

    public void Resolve(T o)
    {
        _taskCompletionSource.TrySetResult(o);
    }

    public void Reject(Error error)
    {
        _taskCompletionSource.TrySetException(error);
    }
}
